### PR TITLE
PR 3: Adding EmojiPicker to the Edit Comment

### PR DESCRIPTION
**Pull request by mananjadhav**
### Details

- This is the third PR to close the tagged issue. It adds the EmojiPicker to the Edit Comment.  Refer the comment https://github.com/Expensify/App/issues/3245#issuecomment-1025441734

### Fixed Issues

$ https://github.com/Expensify/App/issues/3245

### Tests
1. Tested the EmojiPicker on Composer
2. Tested frequently used emojis
3. Tested skintone for emojis
4. Tested searchInput focus for EmojiPicker on Load
5. Tested scroll for EmojiPicker (had issues with earlier implementation)
6. Closing EmojiPicker doesn't shift the focus to the edit comment
7. EmojiPicker positioning switching between Edit comment and Composer

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Go to any chat
2. Select any own message
3. Click on Edit Comment
4. You'll see an Emoji icon and click on the icon
5. Select your emoji, and ensure that it gets entered into the comment.

- [x] Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web

web-edit-emoji-picker

#### Mobile Web

mweb-edit-emoji-picker

#### Desktop

desktop-edit-emoji-picker

#### iOS

ios-edit-emoji-picker

#### Android

android-edit-emoji-picker

### CODE REVIEW

**Comment by mananjadhav** [comment]
@parasharrajat for `ReportActionCompose`, there's this code on `componentDidUpdate`

```
if (this.shouldFocusInputOnScreenFocus && this.props.isFocused
    && prevProps.modal.isVisible && !this.props.modal.isVisible) {
    this.focus();
}
```

It seems you're the code owner. I can't find how `isFocused` as passed as props and need some help in understanding the `props.modal`.

 What is happening is, after I close EmojiPicker on any edit comment, it calls this `componentDidUpdate` of composer and focuses the compose input.

**Comment by parasharrajat** [comment]
Ok. I see. I will tell you what is the use of it and how to manage this when I will review the PR. 

I am waiting on the first PR to be merged. We will go step by step.

File: src/pages/home/report/ReportActionItemMessageEdit.js
Comment (parasharrajat) [inline src/pages/home/report/ReportActionItemMessageEdit.js:198]:
```suggestion
        const shouldDisableEmojiPicker = ReportUtils.chatIncludesConcierge(this.props.report) && User.isBlockedFromConcierge(this.props.blockedFromConcierge) || ReportUtils.isArchivedRoom(this.props.report);
```

File: src/pages/home/report/ReportActionItemMessageEdit.js
Comment (parasharrajat) [inline src/pages/home/report/ReportActionItemMessageEdit.js:222]:
```suggestion
                        isDisabled={shouldDisableEmojiPicker}
```

File: src/pages/home/report/ReportActionItemMessageEdit.js
Comment (parasharrajat) [inline src/pages/home/report/ReportActionItemMessageEdit.js:180]:
```suggestion
```

File: src/pages/home/report/ReportActionItemMessageEdit.js
Comment (parasharrajat) [inline src/pages/home/report/ReportActionItemMessageEdit.js:165]:
```suggestion

```

File: src/pages/home/report/ReportActionItemMessageEdit.js
Comment (parasharrajat) [inline src/pages/home/report/ReportActionItemMessageEdit.js:64]:
```suggestion
```

File: src/pages/home/report/ReportActionItemMessageEdit.js
Comment (parasharrajat) [inline src/pages/home/report/ReportActionItemMessageEdit.js:52]:
```suggestion
```

File: src/pages/home/report/ReportActionItemMessageEdit.js
Comment (parasharrajat) [inline src/pages/home/report/ReportActionItemMessageEdit.js:42]:
```suggestion
        /** Participants associated with current report */
```

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Comment by parasharrajat** [comment]
> Pending: Closing EmojiPicker doesn't shift the focus to the edit comment.

For this, you can use `shouldSetModalVisibility` flag. If you set this to false, the focus is not set to the composer when the modal is closed. 

Now, you can set the focus back to editbox same as   

https://github.com/Expensify/App/blob/36c6f8a2ccb4a846daff664036571c2cdf1fc164/src/pages/home/report/ReportActionItemMessageEdit.js#L134

**Comment by parasharrajat** [comment]
Any update on PR?

**Comment by mananjadhav** [comment]
@parasharrajat @Julesssss PR updated with the feedback and bugs resolved. Can you please review it?

File: src/pages/home/report/ReportActionItemMessageEdit.js
Comment (parasharrajat) [inline src/pages/home/report/ReportActionItemMessageEdit.js:41]:
```suggestion
```

File: src/pages/home/report/ReportActionItemMessageEdit.js
Comment (parasharrajat) [inline src/pages/home/report/ReportActionItemMessageEdit.js:162]:
```suggestion
```

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/components/EmojiPicker/index.js
Comment (mananjadhav) [reply src/components/EmojiPicker/index.js:82]: I had earlier not done it here if you see, I want the picker to be visible after the anchor position is calculated. With the earlier case, I faced this issue https://github.com/Expensify/App/issues/3245#issuecomment-1061629324

Comment (parasharrajat) [reply src/components/EmojiPicker/index.js:82]: Yup. I saw that. 

> But you can use this method to get the dimensions
Use this method to get dimensions only. Then where you are opening the picker and set the dimensions on state. Then create another method that will also get the dimensions from it and update the state. This new method will be used in Dimensions change.

**Comment by parasharrajat** [comment]
Any update @mananjadhav ?

**Comment by mananjadhav** [comment]
Working on it today. Can you confirm the state change is the only change and no other changes?

**Comment by parasharrajat** [comment]
Yup. Rest looks good to me.

File: src/components/EmojiPicker/index.js
Comment (parasharrajat) [inline src/components/EmojiPicker/index.js:33]: Have you tested it after the change? Do you think this method needs binding with this?

Comment (mananjadhav) [reply src/components/EmojiPicker/index.js:34]: Updated

**Comment by parasharrajat** [comment]
@mananjadhav Could you please merge the main? I am testing it now. Sorry for the delay.

**Comment by parasharrajat** [comment]
@shawnborton is the margin correct for picker icon? 
image

**Review by parasharrajat — COMMENTED** [review]
LGTM but awaiting design review.

**Comment by mananjadhav** [comment]
> @mananjadhav Could you please merge the main? I am testing it now. Sorry for the delay.

Done.

**Comment by parasharrajat** [comment]
@Expensify/design is the margin correct for picker icon from edges of Edit box? 
image

**Comment by megankelso** [comment]
that looks a bit too narrow to me, I would increase it some.

**Comment by michelle-thompson** [comment]
Agree with Megan. Also it doesn't look centered through the middle? General rule of thumb, margins on top, bottom, and sides, should look the same.

**Comment by mananjadhav** [comment]
I'll fix the margin-right, in the edit mode, but vertical alignment looks fine as far as I can see. Here's the screenshot for edit and create mode. I used the same style, will 3px more on the right be fine?

Screenshot 2022-03-23 at 11 46 44 PM

**Review by parasharrajat — APPROVED** [review]
LGTM. 

cc: @Julesssss 

#### PR Reviewer Checklist
- [x] I verified the PR has a small number of commits behind `main`
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the testing environment is mentioned in the test steps
- [x] I verified testing steps cover success & fail scenarios (if applicable)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors related to changes in this PR
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified comments were added when the code was not self explanatory
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files (if applicable)
    - [x] I verified proper naming convention for platform-specific files was followed (if applicable)
    - [x] I verified [style guidelines](https://github.com/Expensify/App/blob/main/STYLING.md) were followed
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components are not impacted by changes in this PR (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified the UI performance was not affected (the performance is the same than `main` branch)
- [x] If a new component is created I verified that a similar component doesn't exist in the codebase

:ribbon: :eyes: :ribbon:  C+ reviewed

**Review by Julesssss — APPROVED** [review]

File: src/pages/home/report/ReportActionItemMessageEdit.js
Comment (cubuspl42) [inline src/pages/home/report/ReportActionItemMessageEdit.js:219]: Coming from https://github.com/Expensify/App/issues/30119

If, at the beginning, a composer's emoji picker is open, opening some _other_ emoji picker results in a race:
- Here, we focus the composer because the original emoji picker hides
- The other emoji picker focuses its search field

**Pull request merged**
